README
The Gtk#/Gnome# website can be found at: http://gtk-sharp.sourceforge.net/

Gnome# is a .NET language binding for assorted GNOME libraries.  Gnome# is 
free software, licensed under the GNU LGPL.

Building & Installing Gnome#:
---------------------------

    The build is the traditional:

        ./configure
        make
        make install

    You may want to consider using configure's prefix option to install
    Gnome# using the same prefix as Mono.  That way all of your .NET assemblies
    get placed in the same place, and you don't need to do any extra "configuring"
    to make it so mono can find your assemblies.  In other words,
    doing something like:

        ./bootstrap-<version> --prefix=/the/path/that/was/used/for/mono
        make
        make install

    (Of course, replace "/the/path/that/was/used/for/mono" with whatever path
    which was used for Mono.  This might have been "/usr", "/usr/local", or
    something similar.)

Discussion & Support:
---------------------

    A mailing list for Gnome# discussion is available.

    You can subscribe to the mailing list by visiting:

        http://lists.ximian.com/mailman/listinfo/gtk-sharp-list

    And following the instructions (on that page) to subscribe.
    Messages are posted on this mailing list by sending them to:

        gtk-sharp-list@ximian.com

    (The mailing list requires you to subscribe in order to post
    messages.)

    An archive of this mailing list can be found at:

        http://lists.ximian.com/archives/public/gtk-sharp-list/

    Further, a Wiki is available for Gnome#, and can be found at:

        http://www.nullenvoid.com/gtksharp/wiki/

    Also, people can get help with and discuss Gtk# on IRC via the
    #mono channel on the irc.gnome.org IRC server.

    People looking for general help with C# should visit the
    #c# channel on irc.freenode.net IRC server.


Developers:
-----------

    For developers wishing to "get started" with Gnome#, they are encouraged
    to read the Mono Hand Book:

        http://www.go-mono.com/tutorial

    In the Mono Hand Book, Chapter 21 is on Gtk#.  (In the Mono Hand Book,
    the Gtk# .NET bindings are refered to as GNOME.NET.)


Hackers:
--------

    For those who wish to help with the development of Gnome#, they should
    read the file named: HACKING.

    Also, anyone wishing to hack Gnome# is encouraged to join the Gtk#
    mailing list.  And to visit the #mono IRC channel (on irc.gnome.org).


ChangeLog
2013-01-28 Samuel Gyger <samuel@gyger.at>
        * gconf/GConf/ClientBase.cs: Initialise libdbus' 
        dbus_threads_init_default () to solve Issues with 
        multithreaded applications. 

2010-05-11  Mike Kestner  <mkestner@novell.com>

	* configure.in.in: work around removal of Mono.GetOptions.dll
	in mono trunk (2.8).

2009-03-16  Mike Kestner  <mkestner@novell.com>

	* Makefile.include: parallel make fix.
	* gconf/GConf/Makefile.am: parallel make fix.
	* gconf/GConf.PropertyEditors/Makefile.am: parallel make fix.
	Patch from Peter Alfredsen.  [Fixes #483251]

2009-03-03  Mike Kestner  <mkestner@novell.com>

	* bootstrap-2.24: bump version to 2.24.1 and tag.

2009-01-27  Mike Kestner  <mkestner@novell.com>

	* bootstrap-2.24: add back the POLICY_VERSION entries. Patch from
	Scott Peterson.
	* gnome/Makefile.am: add a POLICY_VERSION variable to override
	the global setting, since gnome-sharp.dll is not compatible with
	previous versions due to the gnomeprint and gnomepanel removal.
	[Fixes #469663]

2008-09-08  Mike Kestner  <mkestner@novell.com>

	* bootstrap-2.24: bump gtk-sharp req to 2.12.2.  Bump version
	to 2.24.0 for release.

2008-08-06  Mike Kestner  <mkestner@novell.com>

	* *: Update to Gnome 2.23.5 release API.  Remove print and panel
	APIs from gnome-sharp.dll into their own assemblies now located in
	gnome-desktop-sharp.

2008-05-14  Stephane Delcroix  <sdelcroix@novell.com>

	* gconf/GConf/gconf-sharp.dll.config.in:
	* gconf/GConf/ClientBase.cs: use GType.Init () instead of
	g_type_init ();

2008-05-14  Stephane Delcroix  <sdelcroix@novell.com>

	* bootstrap-2.20: bump to 2.20.2

	* gconf/GConf/gconf-sharp.dll.config.in:
	* gconf/GConf/ClientBase.cs: call g_type_init during initialization.
	Fixes bnc #389328.

2008-04-21  Mike Kestner  <mkestner@novell.com>

	* bootstrap-2.20: bump version.

2008-04-17  Mike Kestner  <mkestner@novell.com>

	* bootstrap-generic: substitute the GTK_SHARP_REQUIRED_VERSION.
	* bootstrap-2.20: check for gtk 2.12.0 and gtk-sharp 2.12.1.
	* configure.in.in: separate the gtk+ and gtk-sharp version variables.

2008-04-17  Mike Kestner  <mkestner@novell.com>

	* gnome/Program.custom: use GLib.Object.Harden to avoid some
	reflection.

2008-04-17  Mike Kestner  <mkestner@novell.com>

	* bootstrap-2.20: bump gtk-sharp requirement to 2.12.1.
	* gnome/Program.custom: revert QuitPrepare hook and instead do an
	obnoxious hack to disable the toggle_ref mechanism for GnomeProgram.
	Uses reflection since I didn't want to expose this hack in the public
	GLib.Object API.

2008-04-14  Mike Kestner  <mkestner@novell.com>

	* gnome/Program.custom: hook into Gtk.Application.QuitPrepare to
	dispose our managed resources prior to invoking gtk_main_quit.  Avoids
	a crash on exit() in toggle_ref usage after managed releases have been 
	released.

2008-03-04  Mike Kestner  <mkestner@novell.com>

	* bootstrap-2.20: bump version and tag.

2008-02-29  Mike Kestner  <mkestner@novell.com>

	* gnome/gnome-api-2.20.raw: reparse
	* sources/gnome-sharp-2.20-sources.xml: hide private gailcanvas hdrs.
	hide gnome-print-gdi.h.

2008-01-28  Mike Kestner  <mkestner@novell.com>

	* bootstrap-2.20: bump svn version.

2008-01-28  Mike Kestner  <mkestner@novell.com>

	* gnome/Gnome.metadata: hide new Global class which was hiding
	stable API.  [Fixes #356649]

2008-01-11  Mike Kestner  <mkestner@novell.com>

	* configure.in.in: integrate build service patch from Wade.
	* bootstrap-2.20: update to version 2.19.90.

2007-12-12  Mike Kestner  <mkestner@novell.com>

	* * Move to 2.20 API.

2007-12-11  Mike Kestner  <mkestner@novell.com>

	* configure.in.in: move vte, rsvg, and gtkhtml to gnome-desktop-sharp.
	* Makefile.am: ditto
	* sample/Makefile.am: ditto

2007-12-07  Mike Kestner  <mkestner@novell.com>

	* Makefile.include: some refinement of the variable mangling.
	* configure.in.in: expand new pc.in.
	* gconf/GConf/gconf-sharp-2.0.pc.in: remove gnome-sharp Requires and
	peditors.dll from the Libs: var.
	* gconf/GConf.PropertyEditors/Makefile.am: install new pc.
	* gconf/GConf.PropertyEditors/gconf-sharp-peditors-2.0.pc.in: expose
	new pc file since this is conditional and the gconf-sharp.dll build 
	is not.
 
2007-12-07  Wade Berrier  <wberrier@novell.com>

	* gconf/GConf/Makefile.am: Copy AssemblyInfo.cs to the current dir so
	that csc can find it.  (csc doesn't understand ../..)

2007-12-07  Mike Kestner  <mkestner@novell.com>

	* gconf/GConf/Makefile.am: 
	* gconf/GConf.PropertyEditors/Makefile.am: 
	use top_builddir instead of .. to avoid win32 build issues.

2007-12-04  Mike Kestner  <mkestner@novell.com>

	* gconf/GConf/Makefile.am: remove GNOME condition.  It's only needed
	on the peditor build.

2007-10-16  Mike Kestner  <mkestner@novell.com>

	* gnome/CanvasItem.custom: marshal GType as IntPtr.
	[Fixes comment #8 of 327058]

2007-10-11  Mike Kestner  <mkestner@novell.com>

	* gconf/GConf/Value.cs: Deal with zero length arrays gracefully.
	Thanks to Sebastian Dröge <slomo@circular-chaos.org> for a
	patch this one was derived from.  [Fixes #325516]

2007-09-07  Mike Kestner  <mkestner@novell.com>

	* AssembyInfo.cs.in : add [IgnoreClassInitializers]

2007-08-16  Mike Kestner  <mkestner@novell.com>

	* vte/Vte.metadata : rule from Peter Johanson to mark a string
	const.

2007-08-13  Mike Kestner  <mkestner@novell.com>

	* gnome/CanvasPathDef.custom : adapt to generated pinvoke sig change.

2007-07-27  Mike Kestner  <mkestner@novell.com>

	* configure.in.in : add blurb about the gnome-sharp.dll requirements
	since it seems to be a FAQ for people trying to build from sources.

2007-07-03  Mike Kestner  <mkestner@novell.com>

	* bootstrap-generic : use automake --foreign to keep automake-1.10 from
	blowing up on the gnu make function usage.  grrr.
	* sample/gconf/Makefile.am : use install-data-hook to make 1.10 happy.

2007-05-03  Wade Berrier  <wberrier@novell.com>

	* gconf/tools/gconfsharp2-schemagen.in:

gconf/GConf/Value.cs
// Value.cs -
//
// Author: Rachel Hestilow  <hestilow@nullenvoid.com>
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of version 2 of the Lesser GNU General 
// Public License as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this program; if not, write to the
// Free Software Foundation, Inc., 59 Temple Place - Suite 330,
// Boston, MA 02111-1307, USA.

namespace GConf
{
	using System;
	using System.Collections;
	using System.Runtime.InteropServices;

	internal enum ValueType
	{
		Invalid,
		String,
		Int,
		Float,
		Bool,
		Schema,
		List,
		Pair
	};

	internal struct NativeValue
	{
		public ValueType type;
	}

	internal class InvalidValueTypeException : Exception
	{
	}
	
	internal class Value : IDisposable  
	{
		IntPtr Raw = IntPtr.Zero;
		ValueType val_type = ValueType.Invalid;
		bool managed = true;
		
		[DllImport("gconf-2")]
		static extern void gconf_value_set_string (IntPtr value, string data);
		[DllImport("gconf-2")]
		static extern void gconf_value_set_int (IntPtr value, int data);
		[DllImport("gconf-2")]
		static extern void gconf_value_set_float (IntPtr value, double data);
		[DllImport("gconf-2")]
		static extern void gconf_value_set_bool (IntPtr value, bool data);

		ValueType LookupType (System.Type t)
		{
			if (t.IsArray) {
				return ValueType.List;
			} else if (t == typeof (string)) {
				return ValueType.String;
			} else if (t == typeof (int)) {
				return ValueType.Int;
			} else if (t == typeof (double)) {
				return ValueType.Float;
			} else if (t == typeof (bool)) {
				return ValueType.Bool;
			} else if (t == typeof (ICollection)) {
				return ValueType.List;
			} else {
				return ValueType.Invalid;
			}
		}

		public void Set (object data)
		{
			if (data == null)
				throw new NullReferenceException ();
			
			ValueType type = LookupType (data.GetType ());
			Set (data, type);
		}

		[DllImport("gconf-2")]
		static extern IntPtr gconf_value_set_list_nocopy (IntPtr value, IntPtr list);
		
		[DllImport("gconf-2")]
		static extern IntPtr gconf_value_set_list_type (IntPtr value, ValueType vtype);
		
		void Set (object data, ValueType type)
		{
			if (data == null)
				throw new NullReferenceException ();

			switch (type)
			{
				case ValueType.String:
					gconf_value_set_string (Raw, (string) data);
					break;
				case ValueType.Int:
					gconf_value_set_int (Raw, (int) data);
					break;
				case ValueType.Float:
					gconf_value_set_float (Raw, (double) data);
					break;
				case ValueType.Bool:
					gconf_value_set_bool (Raw, (bool) data);
					break;
				case ValueType.List:
					ValueType listType;
					GLib.SList list = GetListFromCollection ((ICollection) data, out listType);
					gconf_value_set_list_type (Raw, listType);
					gconf_value_set_list_nocopy (Raw, list.Handle);
					break;
				default:
					throw new InvalidValueTypeException ();
			}
		}
		
		GLib.SList GetListFromCollection (ICollection data, out ValueType listType)
		{
			object [] arr = (object []) Array.CreateInstance (typeof (object), data.Count);
			data.CopyTo (arr, 0);

			listType = ValueType.Invalid;
			GLib.SList list = new GLib.SList (IntPtr.Zero);
			GC.SuppressFinalize (list);

			if (data.Count == 0) {
				if (data is string [])
					listType = ValueType.String;
				else if (data is int [])
					listType = ValueType.Int;
				else if (data is double [])
					listType = ValueType.Float;
				else if (data is bool [])
					listType = ValueType.Bool;
			}

			if (arr.Length == 0) {
				if (data is Array)
					listType = LookupType (data.GetType ().GetElementType ());
				else
					throw new InvalidValueTypeException ();
				return list;
			}

			foreach (object o in arr) {
				ValueType type = LookupType (o.GetType ());
				if (listType == ValueType.Invalid)
					listType = type;

				if (listType == ValueType.Invalid || type != listType)
					throw new InvalidValueTypeException ();

				Value v = new Value (o);
				GC.SuppressFinalize (v);
				list.Append (v.Raw);
			}
			
			return list;
		}

		[DllImport("gconf-2")]
		static extern IntPtr gconf_value_get_string (IntPtr value);
		
		[DllImport("gconf-2")]
		static extern int gconf_value_get_int (IntPtr value);
		
		[DllImport("gconf-2")]
		static extern double gconf_value_get_float (IntPtr value);
		
		[DllImport("gconf-2")]
		static extern bool gconf_value_get_bool (IntPtr value);
		
		[DllImport("gconf-2")]
		static extern IntPtr gconf_value_get_list (IntPtr value);
		
		public object Get ()
		{
			switch (val_type)
			{
				case ValueType.String:
					return GLib.Marshaller.Utf8PtrToString (gconf_value_get_string (Raw));
				case ValueType.Int:
					return gconf_value_get_int (Raw);
				case ValueType.Float:
					return gconf_value_get_float (Raw);
				case ValueType.Bool:
					return gconf_value_get_bool (Raw);
				case ValueType.List:
					GLib.SList list = new GLib.SList (gconf_value_get_list (Raw), typeof (Value));
					Array result = Array.CreateInstance (GetListType (), list.Count);
					int i = 0;
					foreach (Value v in list) {
						((IList) result) [i] =  v.Get ();
						v.managed = false; // This is the trick to prevent a crash
						i++;
					}

					return result;
				default:
					throw new InvalidValueTypeException ();
			}
		}

		[DllImport("gconf-2")]
		static extern ValueType gconf_value_get_list_type (IntPtr value);

		Type GetListType ()
		{
			ValueType vt = gconf_value_get_list_type (Raw);
			switch (vt) {
			case ValueType.String:
				return typeof (string);
			case ValueType.Int:
				return typeof (int);
			case ValueType.Float:
				return typeof (float);
			case ValueType.Bool:
				return typeof (bool);
			case ValueType.List:
				return typeof (GLib.SList);
			default:
				throw new InvalidValueTypeException ();
			}
		}

		[DllImport("gconf-2")]
		static extern IntPtr gconf_value_new (ValueType type);
		
		public Value (ValueType type)
		{
			Raw = gconf_value_new (type);
		}

		void Initialize (object val, ValueType type)
		{
			Raw = gconf_value_new (type);
			val_type = type;
			Set (val, type);	
		}

		public Value (IntPtr raw)
		{
			Raw = raw;
			NativeValue val = (NativeValue) Marshal.PtrToStructure (raw, typeof (NativeValue));
			val_type = val.type;
		}
/*
		public Value (string val)
		{
			Initialize (val, ValueType.String);
		}

		public Value (int val)
		{
			Initialize (val, ValueType.Int);
		}
		
		public Value (double val)
		{
			Initialize (val, ValueType.Float);
		}

		public Value (bool val)
		{
			Initialize (val, ValueType.Bool);
		}
*/
		public Value (object val)
		{
			Initialize (val, LookupType (val.GetType ()));
		}
		public bool Managed
		{
			get { return managed; }
			set { managed = value; }
		}

		[DllImport("gconf-2")]
		static extern void gconf_value_free (IntPtr value);
		
		~Value ()
		{
			Dispose (false);
		}
		
		public void Dispose ()
		{
			Dispose (true);
			GC.SuppressFinalize (this);
		}

		protected virtual void Dispose (bool disposing)
		{
			if (managed && Raw != IntPtr.Zero)
			{
				gconf_value_free (Raw);
				Raw = IntPtr.Zero;
			}
		}

		public IntPtr Handle
		{
			get {
				return Raw;
			}
		}
	}
}



gconf/GConf/ClientBase.cs
// ClientBase.cs -
//
// Author: Rachel Hestilow  <hestilow@nullenvoid.com>
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of version 2 of the Lesser GNU General 
// Public License as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this program; if not, write to the
// Free Software Foundation, Inc., 59 Temple Place - Suite 330,
// Boston, MA 02111-1307, USA.


namespace GConf
{
	using System;
	using System.Runtime.InteropServices;
	
	public abstract class ClientBase
	{
		public abstract object Get (string key);
		internal abstract void SetValue (string key, Value val);

		public void Set (string key, object val)
		{
			SetValue (key, new Value (val));
		}

		[DllImport("gconf-2")]
		static extern bool gconf_is_initialized ();
	
		[DllImport("gconf-2")]
		static extern bool gconf_init (int argc, IntPtr argv, out IntPtr err);
		
		[DllImport("dbus")]
		static extern Int32 dbus_threads_init_default ();

		internal void Initialize ()
		{
			/* Initialize DBus Glib for multithreading -- It can be called as many times as necessary. */
			dbus_threads_init_default ();
			
			if (!gconf_is_initialized ())
			{
				IntPtr err;
				gconf_init (0, IntPtr.Zero, out err);
				if (err != IntPtr.Zero)
					throw new GLib.GException (err);
				GLib.GType.Init ();
			}
		}
	}
}

